### DISCUSSION TITLE
Recommended procedure to debug code generation for FPGAs from its python interface

### DISCUSSION THREAD

**Discussion by krishnakumarg1984** [spcl/dace#1054] [Q&A] [answered]
For someone with C++ experience, but who is completely  unfamiliar with Dace, what is the recommended route towards debugging dace compiler code generation issues for FPGAs? 

We are using Dace from its python interface, and would like to contribute to upstream through fixes for issues like #1036 with the dace developers and the community's support.

**Comment by tbennun** [comment, marked as answer]
Hi @krishnakumarg1984, thank you for helping contribute to DaCe!

The code generator can be debugged in different ways. If the code generation process itself is failing, the Python debugger would be enough to try and understand what is going on. For compiler errors, like the one in #1036, there are multiple ways of approaching this.

First, the code generation process is relatively straightforward and is visualized in the following gif:

![code generation process](https://spcldace.readthedocs.io/en/docs/_images/codegen.gif)

So if there is a particular hint in the error (array name, kernel) that can pinpoint which node / memlet might be problematic, it would be advisable to place a conditional breakpoint on the function that generates that kind of node, conditioned on the specific label you are looking for. For example, if a map named "kernel_113" in the Xilinx codegen is suspected as the source of the issue, I would place a breakpoint in the `MapEntry` code generator, e.g., [here](https://github.com/spcl/dace/blob/ffade28b99a9b2ef53cad219a153865a601eade5/dace/codegen/targets/fpga.py#L1849).

If it is hard to understand from which line in the code generator the issue has emerged, we created a configuration entry (disabled by default) that adds, to each line in the generated code, the code generator file/line that generated it. It is `compiler.codegen_lineinfo` in `.dace.conf`, or enabled through setting the envvar `DACE_compiler_codegen_lineinfo=1`. It will then create a source map in the appropriate `.dacecache` subfolder (where compilation failed) that you can explore through reading it or through the Visual Studio Code extension.


Good luck! Please feel free to ask any question.
